ethereum/sidecar/server.go

```diff
@@ -951,11 +965,118 @@ func (s *Server) fetchAssetsUnlockConfirmedEvents(
 	return result, nil
 }
 
-//nolint:unparam
-func (s *Server) attestAssetsUnlockedEvents(ctx context.Context) error {
-	// TODO: Implement
-	<-ctx.Done()
-	return nil
+func (s *Server) unqueueAttestation() *bridgetypes.AssetsUnlockedEvent {
+	s.attestationMutex.Lock()
+	defer s.attestationMutex.Unlock()
+	if len(s.attestationQueue) == 0 {
+		return nil
+	}
+
+	attestation := s.attestationQueue[0]
+	s.attestationQueue = s.attestationQueue[1:]
+
+	return &attestation
+}
+
+func (s *Server) attestAssetsUnlockedEvents(ctx context.Context) {
+	ticker := time.NewTicker(1 * time.Second)
+	defer ticker.Stop()
+
+	for {
+		select {
+		case <-ticker.C:
+			for attestation := s.unqueueAttestation(); attestation != nil; attestation = s.unqueueAttestation() {
+				bridgeAssetsUnlocked := &portal.MezoBridgeAssetsUnlocked{
+					UnlockSequenceNumber: attestation.UnlockSequence.BigInt(),
+					Recipient:            attestation.Recipient,
+					Token:                common.HexToAddress(attestation.Token),
+					Amount:               attestation.Amount.BigInt(),
+					Chain:                uint8(attestation.Chain), //nolint:gosec // G115: Chain is known to be within uint8 range
+				}
+
+				ok, err := s.attestationValidator.IsValid(ctx, bridgeAssetsUnlocked)
+				if err != nil {
+					// this is only context cancellation
+					return
+				}
+				if !ok {
+					continue
+				}
+
+				ok, err = s.attestationValidator.IsConfirmed(bridgeAssetsUnlocked)
+				if err != nil {
+					s.logger.Error("couldn't confirm attestation", "attestation", attestation.String(), "error", err)
+				}
+				if ok {
+					s.logger.Info("attestation already confirmed", "attestation", attestation.String())
+					continue
+				}
+
+				delay := s.submissionQueue.GetSubmissionDelay(bridgeAssetsUnlocked)
+
+				s.logger.Info("waiting for attestation submission slot", "delay", delay)
+
+				// wait for our turn to submit
+				select {
+				case <-time.After(delay):
+					s.logger.Info("starting processing attestation", "attestation", attestation.String())
+				case <-ctx.Done():
+					s.logger.Info("topping assets unlocked attestations to context cancellation")
+					return
+				}
+
+				for {
+					ok, err := s.attestationValidator.IsConfirmed(bridgeAssetsUnlocked)
```

Comment:
I think we also need to add a backoff at the begining of this loop to give a chance to reconnect to ETH upon error. Otherwise, we will keep bomb it with requests. Let's of course omit it for the first iteration.

Reply:
48e4fcfe